lib/dynamoid/adapter_plugin/aws_sdk_v2.rb

```diff
@@ -559,6 +569,15 @@ def scan(table_name, scan_hash, select_opts = {})
           record_count = 0
           scan_count = 0
           loop do
+            # Adjust the limit down if the remaining record or scan limit are
+            # lower to obey limits. We can assume the difference won't be
+            # negative due to break statements below.
+            if request[:limit] && record_limit && record_limit - record_count < request[:limit]
+              request[:limit] = record_limit - record_count
+            elsif request[:limit] && scan_limit && scan_limit - scan_count < request[:limit]
+              request[:limit] = scan_limit - scan_count
+            end
+
```

Comment:
I don't think this is the case with `limit: 1` DynamoDB will only return at most 1 record but it doesn't mean it won't underlying scan through 100 unmatched items. I think this type of case arises with `scan_limit` being set but to which case you likely hit `scan_limit` before your `record_limit` is satisfied then.

The only related edge case which I noticed during testing (with DynamoDBLocal) but not sure if it's true in production is that SCAN will only scan through maximum of 1MB of data and then return so if your record sizes are large (Forget maybe 400KB is max record size?) then if you are requesting for records that total 1GB then you potentially are making a lot of HTTP calls to request all the items since each response will contain the maximum 1MB but that's a fault of DynamoDB response system and not really anything we can do about it.

The other case I can think of is making 100 HTTP calls would happen only if there are 100 items and the user requests to `batch(1)` which that's by design of the programming.